File: dataset-fetching/fetch.py
import urllib.request
import json
import datetime
import time
import sys
import os
import zipfile

then = str(int(time.mktime(datetime.date(2000, 1, 1).timetuple())))
now = str(int(datetime.datetime.timestamp(datetime.datetime.now()) * 1000))

if not os.path.exists('stocks'):
    os.makedirs('stocks')

def fetch_symbol(symbol):
    if os.path.exists(f'stocks/{symbol}.json'):
        return False
    price_url = (
        'https://query1.finance.yahoo.com/v7/finance/download/'
        + symbol
        + '?period1='
        + then
        + '&period2='
        + now
        + '&interval=1d&events=history&includeAdjustedClose=true'
    )
    dividend_url = (
        'https://query1.finance.yahoo.com/v7/finance/download/'
        + symbol
        + '?period1='
        + then
        + '&period2='
        + now
        + '&interval=1d&events=div&includeAdjustedClose=true'
    )

    with urllib.request.urlopen(price_url) as response:
        data = response.read().decode()
        data = list(map(lambda x : x.split(','), data.split('\n')))[1:]
        data = list(map(lambda x : ({
            "date": x[0],
            "open": float(x[1]) if x[1] != 'null' else 0,
            "high": float(x[2]) if x[2] != 'null' else 0,
            "low": float(x[3]) if x[3] != 'null' else 0,
            "close": float(x[4]) if x[4] != 'null' else 0,
            "volume": float(x[6]) if x[6] != 'null' else 0,
        }), data))
        with urllib.request.urlopen(dividend_url) as response:
            div = response.read().decode()
            div = list(map(lambda x : x.split(','), div.split('\n')))[1:]
            div_object = {div[i][0]: div[i][1] for i in range(len(div))}
            for entry in data:
                if entry["date"] in div_object:
                    entry["dividend"] = div_object[entry["date"]]
            with open('stocks/' + symbol + '.json', 'w') as file:
                file.write(json.dumps(data))
    return True

# read the fetchlist, and fetch each stock listed
with open('fetchlist.json', 'r') as file:
    fetchlist = json.load(file)
    for entry in fetchlist:
        try:
            if fetch_symbol(entry):
                print(entry)
                # we have to wait more than 1/5 sec bt fetches
                # or else yahoo will block us for DDOSing
                time.sleep(1 / 4)
        except:
            print(f'couldn\'t fetch {entry}!')
        sys.stdout.flush() # so that stdout is updated while running

# write updated metadata for this dataset
with open('metadata.json', 'w') as file:
    file.write(json.dumps({
        "version": 1,
        "fetchDate": str(datetime.datetime.now())
    }))

# zip the fetched stocks and userdata
with zipfile.ZipFile('dataset.zip', 'w', compression=zipfile.ZIP_DEFLATED, compresslevel=9) as z:
    for path, directories, files in os.walk('stocks'):
        for f in files:
            z.write(os.path.join(path, f))
    for path, directories, files in os.walk('other'):
        for f in files:
            z.write(os.path.join(path, f))
    z.write('metadata.json')

File: dataset-fetching/fetch.js
import fetch from 'node-fetch'
import fs from 'fs'

function yahoo (text, dividendsText) {
  const list = text.split(/\r?\n/).slice(1)
  const divs = Object.fromEntries(
    dividendsText.split(/\r?\n/).slice(1).map(line => line.split(','))
  )
  return JSON.stringify(list.map(line => {
    const values = line.split(',')
    const ret = {
      date: (new Date(values[0])).toISOString(),
      open: Number(values[1]),
      high: Number(values[2]),
      low: Number(values[3]),
      close: Number(values[4]),
      // skip adjusted close
      volume: Number(values[6])
    }
    if (divs[values[0]]) { ret.dividend = Number(divs[values[0]]) }
    return ret
  }))
}

async function timer (time) {
  return new Promise(resolve => (
    setTimeout(resolve, time * 1000)
  ))
}

const companies = JSON.parse(fs.readFileSync('fetchlist.json'))
const then = (new Date(2000, 0, 1)).valueOf() / 1000
const now = Date.now()
const folder = 'stocks'
try {fs.mkdirSync(folder)} catch (e) {}

for (const symbol of companies) {
  if (symbol.includes('/')) { continue }
  console.log(symbol)
  let text
  let dividendsText
  try {
    const result = await fetch(`https://query1.finance.yahoo.com/v7/finance/download/${symbol}?period1=${then}&period2=${now}&interval=1d&events=history&includeAdjustedClose=true`)
    text = await result.text()
  } catch (e) {}
  try {
    const dividends = await fetch(`https://query1.finance.yahoo.com/v7/finance/download/${symbol}?period1=${then}&period2=${now}&interval=1d&events=div&includeAdjustedClose=true`)
    dividendsText = await dividends.text()
  } catch (e) {}
  fs.writeFileSync(`${folder}/${symbol}.json`, yahoo(text, dividendsText))
  await timer(0.25)
}
